doc/man/ngrt4nd.1
.TH ngrt4nd 1  "Last Update: October 03, 2012" "Version: 1.1.0" "ADMIN COMMANDS"
.SH NAME
ngrt4nd \- The utility to manage the daemon module of NGRT4N.
.SH SYNOPSIS
.B ngrt4nd
[OPTIONS]
.SH DESCRIPTION
This program allows to manage the daemon module of NGRT4N. 
.PP
It sits on the same server than Nagios.
.SH OPTIONS
.TP
\-h
Print a brief help.
.TP
\-c </path/to/status.dat>
Set an alternative location for the Nagios status file. By default, the program looks at /usr/local/var/nagios/status.dat. 
.TP
\-D
Run ngrt4nd in the foreground. This option is useful for debuging, since error and debug messages are printed to the terminal.
.TP
\-P 
Set the authentication token.
.TP
\-T
Print the authentication token.
.TP
\-p <port>
Set the listening port. Default is 1983.
.SH EXAMPLES
.TP
$ ngrt4nd -p 2345
.SH EXIT STATUS
ngrt4nd returns a zero exit status on success. Non zero is returned in case of failure.
.SH AUTHOR
This page has been written by Rodrigue Chakode <rodrigue.chakode@ngrt4n.com>
.SH COPYRIGHT
Copyright (c) 2010-2012 by NGRT4N Project. All rights reserved.
.PP
Visit http://ngrt4n.com for more information.
.SH SEE ALSO
.BR ngrt4n(1)



ChangeLog
###############################
RealOpInsight ChangeLog
###############################
3.0.0b1 (Eliana, b1)
  * Huge refactoring and code simplification
  * Port to Qt 5.x
  * Livestatus is now the default mode selected for retrieving monitoring data
    when setting Nagios-based sources. The use of ngrt4nd needs to be explicitly
    enabled
  * Fixed ticket #25 enabling the ability to cancel the basic authentication for
    Nagios
  * Fixed close action on preference window (ticket #24)
  * Moved piechart from the map to lelf side
  * Improved charting rendering: cleanup and new algorithm
  * Changed default operations console size to 1024 x 768
  * Fixed ticket #27 when events are not well propagated
  * IMPORTANT: implicit check name for ping is not longer supported, the user
    is now required to set that. E.g. mysql-server/ping.

2.4.0 (Federator) - 08/09/2013
  * Fixed ticket #22 : Operations Console doesn't update the status of triggers with Zabbix API v1.4 (Zabbix 2.0.x, x<5)
  * Fixed close action on the preference window (ticket #24)

2.4.0 (Federator) - 08/09/2013
  * Fixed compilation issue on OS X related to basename and QString (ticket #20)
  * Fixed install script for OS X (ticket #21): removed the use of lscpu
    and added option to set the number of CPU
  * Improved install script enabling user mode installation
  * comes with ngrt4n-d4n that brings support for ZeroMQ 3.x

2.4.0beta2 (Federator,b2) - 16/08/2013
  * Fixed ticket #16: adding support for realm authentication within the browser
  * Fixed ticket #17: file path missing in the title bar of the console
  * Fixed ticket #18: Added the possibility to select the source to load within the browser
  * Added logo and misc icons for Nagios, Zenoss, Zenoss, and hierarchy

2.4.0beta1 (Federator,b1) - 08/08/2013
  * Improved support for overriding SSL peer verification
  * Added support for multisources management, up to 10 sources supported
  * Lot of refactoring, imporving loading and updating
  * Added splash screen when loading configuration
  * Added support for dragging and scrolling from the map
  * Added workaround for ticket #13 adding option to override host verification
  * Fixed ticket #14 related to chart panel repositionning
  * Password manager is not longer execute through the configuration wizard

2.3.1 (Briator, p1) - 23/06/2013
  * Added support for overriding SSL peer verification
  * Private release no sufficiently tested for production

2.3.0 (Briator) - 16/05/2013
  * Minor changes from the beta version.
  * Binaries are available for Fedora 17 and 18, openSUSE 12.1, 12.2 and 12.3.
    From this version, we don't longer provide binaries for Fedora 16

2.3.0beta1 (Briator, b1) - 04/04/2013
--------------
  * Added support for MK Livestatus remote socket
  * Added new menus for handling full screen, filtering trouble messages, and
    for changing font size from the operations console
  * Added support for copy/paste as well as more menu shortcuts within the editor
  * Added new icons
  * Added french translation
  * added validator for the Port field
  * Make the IDs of data points case insensitive
  * Minor bug fixes

2.2.1 (Lucia, p1) - 10/03/2013
-------------
  * Added support for Zabbix 1.8.x. Fixed ticket #11
  * Minor upates on metadata handling

2.2.0 (Lucia) - 02/03/2013
--------------
  * Fixed compilation problems on nullptr. See Ticket #8
  * Fixed metadata handling. See ticket #9

2.2.0beta1(Lucia, b1) - 02/02/2013
--------------
  * Added support for Zenoss. Has been tested with Zenoss 4.x but would work
    with Zenoss 3.x
  * Update severity model introducing a fine-grained manageability of incidents
    (UNKNOWN, CRITICAL, MAJOR, MINOR, NORMAL)
  * Added System Tray Icon for notification the main window is inactive
  * Support some Zabbix's meta tag (e.g. {HOST.NAME}) in alarm messages
  * Brings new and optimized data structures and algorithms for better
    performances and a reduced memory footprint
  * Added buttons for back, forward and stop actions through the embedded browser
  * Added support for zeromq 3.x (ticket #3)
  * Added man pages for the commands ngrt4n-manager, ngrt4n-oc and ngrt4n-editor
  * Added contextual help for editor fields
  * Fixed compilation problems on c++11 (tickets #4 and #5)
  * Fixed bug #6 when the Operations Console crashes when the colon is ommited
    on the authchain for Zabbix
  * Changed tooltip font

2.1.0 (M-Sylvie) - 28/11/2013
--------------
  * Added option for hiding/showing the auth chain
  * Fixed bug #1: The severities of Zabbix triggers are now supported.
  * Updated the preferences window for easing the setting of monitoring settings

2.1.0beta1 - 27/10/2012
--------------
  * Added support for Zabbix: Requires Zabbix >= 1.7. 
    Known Issues: the severities of triggers are not yet taken into account. 
    All problems are simply reported. This will be fixed before the stable 
    release (ticket #1). 

  * Released the UI Module and the Daemon Module as seperate packages. 
    The UI Module became 'RealOpInsight' while the Daemon Module, specific 
    to Nagios, is now called 'Daemon Service for Nagios'.

  * VERY IMPORTANT: To support both Zabbix and Nagios without additional 
    configuration constraints for the administrators, the format of configuration 
    files has been modified to detect automatically the target monitoring system 
    at the loading time. Nevertheless, for backward compatibility, the former 
    format is still supported and detected as compatible with Nagios. 
    It's recommended to upgrade the format of your former configuration files: just 
    open the each file from the Editor and resave it (File -> Save).

  * Changed of the format of message exchanges between the Daemon 
    Service for Nagios and the Operations Consoles for better 
    performance and better performance and more flexibility. It's 
    required to upgrade the Daemon Service towards the version 1.1.0.


src/server/ngrt4nd.cpp
/*
 * ngrt4nd.cpp
# ------------------------------------------------------------------------ #
# Copyright (c) 2010-2012 Rodrigue Chakode (rodrigue.chakode@ngrt4n.com)   #
# Last Update : 24-05-2012                                                 #
#                                                                          #
# This file is part of NGRT4N (http://ngrt4n.com).                         #
#                                                                          #
# NGRT4N is free software: you can redistribute it and/or modify           #
# it under the terms of the GNU General Public License as published by     #
# the Free Software Foundation, either version 3 of the License, or        #
# (at your option) any later version.                                      #
#                                                                          #
# NGRT4N is distributed in the hope that it will be useful,                #
# but WITHOUT ANY WARRANTY; without even the implied warranty of           #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            #
# GNU General Public License for more details.                             #
#                                                                          #
# You should have received a copy of the GNU General Public License        #
# along with NGRT4N.  If not, see <http://www.gnu.org/licenses/>.          #
#--------------------------------------------------------------------------#
 */

#include "config.h"
#include "core/ns.hpp"
#include "core/MonitorBroker.hpp"
#include "core/ZmqSocket.hpp"
#include <zmq.h>
#include <cassert>
#include <stdexcept>
#include <iostream>
#include <unistd.h>
#include <crypt.h>
#include <sstream>
#include <fstream>
#include <libgen.h>
#include <iostream>
#include <stdio.h>
#include <memory>


std::string packageName = PACKAGE_NAME;
std::string packageVersion = PACKAGE_VERSION;
std::string packageUrl = PACKAGE_URL;
std::string statusFile = "/usr/local/nagios/var/status.dat";
std::string progName = "";
std::string authChain= "";

std::string help() {
  std::ostringstream msg("SYNOPSIS\n"
                         "	" + progName +" [OPTIONS]\n"
                         "\n"
                         "OPTIONS\n"
                         "	-c FILE\n"
                         "	 Specify the path of the Nagios status file. Default is " + statusFile + ".\n"
                         "	-D\n"
                         "	 Run the program in foreground mode. \n"
                         "	-p\n"
                         "	 Set the listening port. Default is 1983.\n"
                         "	-P\n"
                         "	 Change the authentication token.\n"
                         "	-T\n"
                         "	 Print the authentication token.\n"
                         "	-v\n"
                         "	 Print the version and copyright information.\n"
                         "	-h\n"
                         "	 Print this help.\n\n");
  return msg.str();
}

std::string getVersionMsg(const std::string& progName)
{
  char msg[1024];
  sprintf(msg, "> %s (%s) version %s"
          "\n>> Belongs to NGRT4N Dashboard Monitoring Suite"
          "\n>> Copyright (c) 2010-2013 NGRT4N Project. All rights reserved"
          "\n>> License GNU GPLv3 or later <http://gnu.org/licenses/gpl.html>"
          "\n>> For bug reporting instructions, see: <%s>",
          packageName.c_str(), progName.c_str(), packageVersion.c_str(), packageUrl.c_str());
  return std::string(msg);
}

void ngrt4n::setPassChain(char* authChain)
{
  std::ofstream ofpass;
  ofpass.open(ngrt4n::AUTH_FILE.c_str());
  if(!ofpass.good()) {
      std::clog << "Unable to set the authentication token." << "\n";
      exit(1);
    }
  ofpass << crypt(authChain, salt.c_str());
  ofpass.close();
}

std::string ngrt4n::getPassChain()
{
  std::string authChain;
  std::ifstream pfile;
  pfile.open (ngrt4n::AUTH_FILE.c_str());
  if(!pfile.good()) {
      std::clog << "Unable to load the application's settings" << "\n";
      exit(1);
    }

  pfile >> authChain;
  pfile.close();
  return authChain;
}

int main(int argc, char ** argv)
{
  progName = basename(argv[0]);
  int port = MonitorBroker::DefaultPort;
  bool foreground = false;
  char opt;
  static const char *shotOpt="DTPhvc:p:";
  while ((opt = getopt(argc, argv, shotOpt)) != -1) {
      switch (opt){
        case 'D':
          foreground = true;
          break;
        case 'c':
          statusFile = optarg;
          break;
        case 'p':
          port = atoi(optarg);
          if(port <= 0 ) {
              std::cerr << "Bad port number\n";
              exit(1);
            }
          break;
        case 'P':
          {
            ngrt4n::checkUser();
            ngrt4n::initApp();
            char* pass = getpass("Type a passphrase:");
            ngrt4n::setPassChain(pass);
            std::cout << ngrt4n::getPassChain()<<"\n";
            exit(0);
          }
        case 'T':
          ngrt4n::checkUser();
          std::cout << ngrt4n::getPassChain()<<"\n";
          exit(0);
        case 'v':
          std::cout << getVersionMsg(progName)<<"\n";
          exit(0);
        case 'h':
          std::cout << help();
          exit(0);
        default:
          std::cout << help();
          exit(1);
        }
    }
  ngrt4n::checkUser();
  ngrt4n::initApp();
  authChain = ngrt4n::getPassChain();
  std::clog << getVersionMsg(progName)<< "\nStarting...\n";
  if(!foreground) {
      pid_t pid = fork();
      if(pid <= -1) {
          std::clog << "Failed while starting the daemon service\n";
          exit(1);
        } else if(pid > 0) {
          exit (0);
        }
      setsid();
    }

  std::ostringstream uri;
  uri << "tcp://0.0.0.0:" << port;
  ZmqSocket socket(ZMQ_REP);
  if(!socket.bind(uri.str())) {
      std::clog << "ERROR\n";exit(1);
    }
  std::clog << "Listening address => "<<uri.str()
            << "\nNagios status file => "<<statusFile
            << "\n============>started\n";

  auto monitor = std::unique_ptr<MonitorBroker>(new MonitorBroker(statusFile));
  while (true) {
      std::string msg = socket.recv();
      std::string reply;
      if(msg == "PING") {
          reply = "ALIVE:"+packageVersion;
        } else {
          size_t pos = msg.find(":");
          std::string pass = "";
          std::string sid = "";
          if(pos != std::string::npos) {
              pass = msg.substr(0, pos);
              sid = msg.substr(pos+1, std::string::npos);
            }
          if(pass == authChain) {
              reply = monitor->getInfOfService(sid);
            } else {
              reply = "{\"return_code\" : \"-2\", \"message\" : \"Authentication failed\"}";
            }
        }
      socket.send(reply);
    }
  monitor.reset(NULL);
  return 0;
}
